### DISCUSSION TITLE
Project Update Q1 2025

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#289] [Announcements]
We are well overdue for an update for Q4 so I'm rolling this up into a Q1 2025 update. Anyone who has been following the project, especially via our Discord feed would know that there has been a LOT of activity over the Q4 2024 period.

The biggest thing that has happened is that we now have some regular contributors on board who are really helping to keep the momentum of the project moving forwards. This has been a massive boost for the project not only in helping to keep things moving, but also turning this into a true community project, which was alway the goal. Validation and bug detection is now a fluid process which has exponentially increased what we are able to achieve in any given time frame.

## Current code status

We currently have stable functioning code that is returning reasonable results. We are currently working on validation to both determine and improve bench accuracy and repeatability.

The code is certainly at a stage where we are happy to recommend widespread early-adoption. Whilst we are still working towards the official V2 release, the basic functionality of displaying flow is working to an acceptable level and we are continually working towards increasing accuracy and repeatability. Whilst we are not in a position to quote cfm accuracy figures (simply as we do not have enough data at this point in time), using DIYFB as intended - as a tool for change validation, will work as well as any regular flow bench. 

A quotable CFM accuracy is one of the goals of the V2 release and is one that we are actively working towards.

## V2 PCB Shield

The V2 PCB shield validation is complete, and whilst I have not had the time to prepare and organise sales my plan is to get that ready over the xmas break so that anyone who wants a PCB can place an order in the new year.  There are a couple of minor issues with the PCB but there are [simple workarounds](https://github.com/DeeEmm/DIY-Flow-Bench/discussions/171) for those issues.

## General Focus

The big drive over this period has been to focus on the next release candidate and close out all minor bugs and features. The mantra has been that if it is currently part of the GUI, then it should function correctly. So there has been a feature lock and everyone has been working hard to make sure that all [bugs and inconsistencies](https://github.com/DeeEmm/DIY-Flow-Bench/issues?q=is%3Aopen+is%3Aissue+milestone%3ADIYFB-V.2.0-RC.9) are picked up and addressed.

So whilst the issues list for RC9 ebbs and wanes, hopefully at some point in the near future everything will be closed out and the next release candidate will be published.

## Management of change

In amongst the code changes, and as a result of the extra contributors, we have had some small changes to the change management philosophy. This is really to help manage multiple simultaneous contributions and ensure that the integrity of the main codebase is maintained. This is easy to do when you are the sole contributor, but requires a more considered approach when multiple people are working on the code.

- Current up to date code is now available on the [DEV Branch](https://github.com/DeeEmm/DIY-Flow-Bench/tree/DEV) 
- The master branch is now level with the most recent release.

Changes are now developed on their own dedicated branches based on the DEV branch, and are generally tied to an issue. This compartmentalises each change and ensures that Pull Requests are small and manageable.

When a release is published, the Dev branch is merged with the master branch and the release generated from the master branch code. This ensures that anyone wanting to use the current code can grab it from either the master branch or the release page.

## Firmware updates

One big change to the code is how installation and updates are managed. It is appreciated that for many people using a development environment like the VSCode IDE or the Arduino IDE has a very steep learning curve and so the process of initially setting up your ESP32 and performing future updates has been streamlined. It is no no longer necessary to use an IDE to upload the initial firmware, this can be done using your browser. We are also working on a companion app that will make the process even simpler. 

Once you are up and running firmware updates are taken care of within the GUI and is simple as clicking a button on the configuration settings tab. This should hopefully remove the barrier that many face for adopting projects like DIYFB.

## The future

Any review is of course incomplete without some view of what the project looks like moving forwards. The main immediate goal is get the RC9 release finalised and tested which should definitely happen within Q1 2025. Fingers crossed this is the last release candidate for the V2 release. In fact we have formalised the RC9 release management and have been working hard to focus only on those items.

Once the V2 release is out in the wild then we will be looking at prioritising features for the V2.1 release. The change management goal moving forwards will be to implement a regular release cycle were we can pick a feature, develop a solution, test and then release a version update. It is unclear what period the release cycle will follow at this stage but I suspect that aiming for a quarterly release cycle is likely most sustainable with the possibility for interim releases as they arise or are required.

I have a good feeling about the current state of the project and what it looks like moving into 2025. I want to give a big shout out to everyone who has contributed to the project so far. Thank you. 

Merry Christmas Everyone. 😄 🎄 🎅 

/DM

**Comment by ModchipMan** [comment]
Hi guysI have been following the project via the emails..Hopefully in 2025 I can get back to this project with you all..Merry Xmas guys and have a safe holidays Tony Donn..Sent from my iPhoneOn 12 Dec 2024, at 11:17 am, Mick Percy ***@***.***> wrote:﻿
We are well overdue for an update for Q4 so I'm rolling this up into a Q1 2025 update. Anyone who has been following the project, especially via our Discord feed would know that there has been a LOT of activity over the Q4 2024 period.
The biggest thing that has happened is that we now have some regular contributors on board who are really helping to keep the momentum of the project moving forwards. This has been a massive boost for the project not only in helping to keep things moving, but also turning this into a true community project, which was alway the goal. Validation and bug detection is now a fluid process which has exponentially increased what we are able to achieve in any given time frame.
Current code status
We currently have stable functioning code that is returning reasonable results. We are currently working on validation to both determine and improve bench accuracy and repeatability.
The code is certainly at a stage where we are happy to recommend widespread early-adoption. Whilst we are still working towards the official V2 release, the basic functionality of displaying flow is working to an acceptable level and we are continually working towards increasing accuracy and repeatability. Whilst we are not in a position to quote cfm accuracy figures (simply as we do not have enough data at this point in time), using DIYFB as intended - as a tool for change validation, will work as well as any regular flow bench.
A quotable CFM accuracy is one of the goals of the V2 release and is one that we are actively working towards.
V2 PCB Shield
The V2 PCB shield validation is complete, and whilst I have not had the time to prepare and organise sales my plan is to get that ready over the xmas break so that anyone who wants a PCB can place an order in the new year.  There are a couple of minor issues with the PCB but there are simple workarounds for those issues.
General Focus
The big drive over this period has been to focus on the next release candidate and close out all minor bugs and features. The mantra has been that if it is currently part of the GUI, then it should function correctly. So there has been a feature lock and everyone has been working hard to make sure that all bugs and inconsistencies are picked up and addressed.
So whilst the issues list for RC9 ebbs and wanes, hopefully at some point in the near future everything will be closed out and the next release candidate will be published.
Management of change
In amongst the code changes, and as a result of the extra contributors, we have had some small changes to the change management philosophy. This is really to help manage multiple simultaneous contributions and ensure that the integrity of the main codebase is maintained. This is easy to do when you are the sole contributor, but requires a more considered approach when multiple people are working on the code.

Current up to date code is now available on the DEV Branch
The master branch is now level with the most recent release.

Changes are now developed on their own dedicated branches based on the DEV branch, and are generally tied to an issue. This compartmentalises each change and ensures that Pull Requests are small and manageable.
When a release is published, the Dev branch is merged with the master branch and the release generated from the master branch code. This ensures that anyone wanting to use the current code can grab it from either the master branch or the release page.
Firmware updates
One big change to the code is how installation and updates are managed. It is appreciated that for many people using a development environment like the VSCode IDE or the Arduino IDE has a very steep learning curve and so the process of initially setting up your ESP32 and performing future updates has been streamlined. It is no no longer necessary to use an IDE to upload the initial firmware, this can be done using your browser. We are also working on a companion app that will make the process even simpler.
Once you are up and running firmware updates are taken care of within the GUI and is simple as clicking a button on the configuration settings tab. This should hopefully remove the barrier that many face for adopting projects like DIYFB.
The future
Any review is of course incomplete without some view of what the project looks like moving forwards. The main immediate goal is get the RC9 release finalised and tested which should definitely happen within Q1 2025. Fingers crossed this is the last release candidate for the V2 release. In fact we have formalised the RC9 release management and have been working hard to focus only on those items.
Once the V2 release is out in the wild then we will be looking at prioritising features for the V2.1 release. The change management goal moving forwards will be to implement a regular release cycle were we can pick a feature, develop a solution, test and then release a version update. It is unclear what period the release cycle will follow at this stage but I suspect that aiming for a quarterly release cycle is likely most sustainable with the possibility for interim releases as they arise or are required.
I have a good feeling about the current state of the project and what it looks like moving into 2025. I want to give a big shout out to everyone who has contributed to the project so far. Thank you.
Merry Christmas Everyone. 😄 🎄 🎅
/DM

**Comment by black-top** [comment]
Hi, 
A lot of changes since last time  checked. is there are usable version of code or all still under development and fixing? Is it far from RC9?
In the near future I can do measurments with Siemens 5WK9605  MAF, just need to make adapters for it. The data we have for it is from original bmw m54b30 ecu, I have comparaded the data some time ago.

**Reply by DeeEmm** [reply]
Usable version is under DEV branch

Code level is pretty much RC9. The only thing that is stopping me from releasing is that I need to do further MAF conversion / validation. Just been a bit flat out to do that this week.

You should be able to update using the binary in the [release folder](https://github.com/DeeEmm/DIY-Flow-Bench/tree/DEV/ESP32/DIY-Flow-Bench/release) of the repo. However Depending on what version you previously had you will most likely have to reconfigure your I/O.

Moving forwards I/O management should not change any further, so once you set your I/O in the GUI, the values stored in NVM should not be impacted by future updates, however there are some plans to improve I/O overrides as discussed elsewhere.

There is an official Siemens 5WK99605 thread [here](https://github.com/DeeEmm/DIY-Flow-Bench/discussions/141) Not sure if the data you have is the same.

Will endeavor to get the additional MAF sensors added this week.

**Reply by DeeEmm** [reply]
Before you upgrade. I should add that there's only currently only the following MAF sensors converted...

<img width="392" alt="Screenshot 2025-02-02 at 12 04 40 pm" src="https://github.com/user-attachments/assets/29d513d5-ce5f-47a1-b3b6-e805d811c3b3" />

**Reply by DeeEmm** [reply]
... and I'm only really confident about the data for the Bosch 0280218067. Hence the need for review.

Some of these data sources were collected a long time ago, so I want to make sure that they are valid. 

I did also add a check to the info popup that states the current status of the MAF sensor...

<img width="540" alt="Screenshot 2025-02-02 at 12 07 45 pm" src="https://github.com/user-attachments/assets/384db0c7-bf46-4c6d-a915-425c0941d484" />

If anyone does use these unvalidated / untested sensors and verifies they are okay, I will update the status. The info link takes you directly to the discussion page in the forum for the current MAF sensor.

**Reply by black-top** [reply]
Yes, my current setup is based on the 0280218067, same as before.  Don't remember the exact version, but it will be old because it was before new maf algorithm implementation. If I upload the current version it should not damage any components before I/O is configured? or it is better to disconnect esp32 from PCB? Btw, my pressure sensor is MPXV7025, is available to select at current version?

**Reply by DeeEmm** [reply]
I will make a change to the default settings to disable all I/O. Currently it is set to default (wemos) settings. I think that it is better for everything to be disabled initially and then the end user can enable what is needed.

Separately I will work on the ability to upload a configuration file that can set all I/O in accordance with a specific board. But this might not be until V3.

**Reply by DeeEmm** [reply]
OK. default pin value is now changed to '-1' in DEV branch.

**Reply by black-top** [reply]
The upload process from browser with ESP tool is same as before?

**Reply by DeeEmm** [reply]
Yes. If you have a Mac the firmware uploader is also now working. There's a link to it in the firmware uploader thread - https://github.com/DeeEmm/DIY-Flow-Bench/discussions/343

**Reply by black-top** [reply]
ok, thanks.

**Comment by DeeEmm** [comment]
I've added the additional Bosch sensors from the HFM5 datasheet. This includes the following part numbers

```
        "BOSCH 0280002421"
        "BOSCH 0280218008"
        "BOSCH 0280217531"
        "BOSCH 0280218019"
        "BOSCH 0280217123"
```

I've also tested the generated coefficients and calculated output against the data tables in the datasheet and the results returned are good. So I've marked these as 'Tested'. Note that this doesn't mean 'validated'.

Same applies for the Other sensors apart from the ACDELCO 92281162. The results returned align with the data we have, but as the provenance of the source is unvalidated the values should be treated with caution.


### RC9

So this was the last aspect that I wanted to sort out before the RC9 release.

Once I've received a bit of feedback, provided that everything is good I will package it up and push RC9 out.

**Comment by DeeEmm** [comment]
These are the sensors currently supported.

<img width="276" alt="Screenshot 2025-02-04 at 6 06 41 pm" src="https://github.com/user-attachments/assets/5cd3cf07-04bb-486d-bab2-fad2514b3b10" />

If your sensor is not listed let me know.

**Comment by DeeEmm** [comment]
(More data on each sensor can be found in the dedicated discussion thread for each sensor)

**Comment by DeeEmm** [comment]
Just went to update the MAF data table in the Wiki on the [component selection](https://github.com/DeeEmm/DIY-Flow-Bench/wiki/Component-selection) page and note there are some additional sensors listed. Will try and validate the data and include it in the code.

Other than that I'm just working through the WIKI to make sure that everything is up to date.

**Reply by black-top** [reply]
I get an error to select pin numbers up to 39, but the pin list at the wiki is up to 99.

**Comment by black-top** [comment]
I uploaded latest DEV and  now I need to set the pins and other hardware configuration.  Is this correct pin mapping from wiki for diyfwb board? A lot of sesnors are set to 99 . It is probably because of i2c?

>  "BOARD_TYPE" : "WEMOS_D1_R32"
 "VAC_SPEED_PIN" : 25 
 "VAC_BLEED_VALVE_PIN" : 26
 "VAC_BANK_1_PIN" : 13
 "VAC_BANK_2_PIN" :12
 "VAC_BANK_3_PIN" :14
 "AVO_STEP_PIN" :  32
 "AVO_DIR_PIN" : 33
 "FLOW_VALVE_STEP_PIN" : 27
 "FLOW_VALVE_DIR_PIN" : 4
 "VCC_3V3_PIN" : 99
 "VCC_5V_PIN" : 2
 "SPEED_SENSOR_PIN" : 15
 "SWIRL_ENCODER_PIN_A" : 99
 "SWIRL_ENCODER_PIN_B" : 99
 "ORIFICE_BCD_BIT1_PIN" : 34
 "ORIFICE_BCD_BIT2_PIN" : 36
 "ORIFICE_BCD_BIT3_PIN" : 39
 "MAF_PIN" : 99
 "REF_PRESSURE_PIN" : 99
 "DIFF_PRESSURE_PIN" : 99
 "PITOT_PIN" : 99
 "TEMPERATURE_PIN" : 99
 "REF_BARO_PIN" : 99
 "HUMIDITY_PIN" : 99
 "SERIAL2_TX_PIN" :16
 "SERIAL2_RX_PIN" :17
 "SDA_PIN" : 21
 "SCL_PIN" : 22
 "SD_CS_PIN" : 5
 "SD_MOSI_PIN" : 23
 "SD_MISO_PIN" : 19             
 "SD_SCK_PIN" : 18
 "WEMOS_SPARE_PIN_1": 99
 

>

**Reply by DeeEmm** [reply]
All pins should now be disabled by default, but I note that I gave them the value of 99. It should be -1. I will correct this.

**Reply by DeeEmm** [reply]
Actually just checked. They are all set to -1 by default.

The 99 value previously used to disable the pin is now -1.

You should see all of the pins status set to -1.

This is what I have set for the DIYFB shield...

<img width="1288" alt="Screenshot 2025-02-05 at 8 36 44 pm" src="https://github.com/user-attachments/assets/7afa0c0f-724a-4ed4-9828-e95b4955ccea" />

**Reply by DeeEmm** [reply]
I will update the Wiki. 👍

**Reply by DeeEmm** [reply]
ok wiki updated. I also added a link to the new 'pins page' this is where the info icons link to from the GUI. it includes a brief description of each pins function

**Comment by black-top** [comment]
Thanks, few more questions.
 What is the correct way for maf trim pot adjustments in this version? With unconnected maf sensor, there is 0.05-0.06V offset, so I should adjust the trim pot until I get 0 Flow value? Or it is better to do the same with connected maf? Or I should adjust the trim pot to mach orifice flow? 
![issue1](https://github.com/user-attachments/assets/9a5787f9-6ec1-4154-b73b-77cbd12cc60e)

**Reply by DeeEmm** [reply]
At zero flow the sensor should be returning zero. This needs to be done with the sensor in circuit not floating.

**Reply by DeeEmm** [reply]
- connect the sensor
- check the MAF voltage value with the bench stopped
- use this value as the trimpot value to bring the baseline voltage reading to zero

Points to note are that the voltage value is the raw value direct from the sensor, whereas the flow values may have corrections applied.

for example the mass flow value may be transposed if a non-standard housing diameter is entered.

**Reply by black-top** [reply]
Just to double check for the calibration, the BOSCH_0280218067 0 to 1V measures reverse flow and above 1V direct flow, with connected maf and no flow I see 1.02V and 11.83kgh flow, so I should adjust the trim pot for 0.02V offset,  or adjust until I get 0 flow value, but in that case reverse flow measurement will be ignored?

**Reply by DeeEmm** [reply]
OK so just checking the code for MPX4250. The sensor has been removed as it is not really suitable for this application. It does not have the needed resolution. This is why it is not working. It's presence in the dropdown has been overlooked. It should have been removed.

For the 250kPa full scale value / 32767 ADC data points the actual usable portion (28"/H2O) or 7kPa only falls over 917 data points. If you're not using a full 28" depression that value is even smaller. This effectively reduces the resolution of the Reference pressure sensor from 16bits  to below that of a 10 bit system. If you're only testing to 10" that can be further reduces to an 8 bit system

The result of this is that the reference pressure resolution is fairly 'notchy' and will cause a lot of jumping and instability of the display.

The MPX4250 is really a bit of a throwback to the very early days of development before things like optimum sensor choice was still to be worked out. I really advise against using it.

**Reply by DeeEmm** [reply]
> Just to double check for the calibration, the BOSCH_0280218067 0 to 1V measures reverse flow and above 1V direct flow, with connected maf and no flow I see 1.02V and 11.83kgh flow, so I should adjust the trim pot for 0.02V offset, or adjust until I get 0 flow value, but in that case reverse flow measurement will be ignored?

The BOSCH_0280218067 does include the reverse flow, so it should be putting out 1v at rest. You will need to adjust for 1v value.

**Reply by black-top** [reply]
Yes, I have used mpx4250 since the beginning of this project, I bought it on purpose because my pump easy overloads even mpx4250 at smaller lift and I want to do testing at much higher pressures than 28"' and also all my previous testing was done with it and only maf readings were given the problems. So it will be good to keep it available in the official code

**Reply by DeeEmm** [reply]
I too initially started off with MPX4250 as I also use them in ECU builds (it is commonly used as a MAP sensor).

The fluctuation in reference pressure will directly impact the MAF readings as the CFM is calculated using the reference pressure. The two are directly connected. The fluctuations you saw were likely due to this issue. I recall we discussed this before. This resolution issue will also impact depression regulation when we move to a modulated system. It is less than optimal and depression control will suffer as a result.

There are much better choices, even if you are running a higher depression. There is no way that your bench will get anywhere near to 250kPA (1000"H2O).

IMHO The drawbacks far outweigh any perceived advantages. The accuracy is of a much higher importance.

The issue with over-pressuring at lower lifts is simply one of procedure. It is the same issue with a regular bench. It is very easy to suck the fluid out of the manometer 😆 

Always start the bench with the bypass valve open and you will never have an issue

**Reply by DeeEmm** [reply]
Are you only using it for pRef?

I really do not want to add it back in, but will add it in until the V1 release. This will give you some time to source a more appropriate sensor.

The main reason that I don't want to include support for this sensor, is that if it is there, people will choose to use it, as it is cheaper than sourcing the proper sensors. The issue is that this then causes accuracy and stability issues which are bad for both project reputation and will burn up a lot of time with support.

**Reply by DeeEmm** [reply]
in V1 I will simply comment it out in the HTML - so if you decide that you must keep using it, you can uncomment it.

However i really strongly advise you not to use this sensor. it is not suitable. Nor will it be supported.

**Reply by black-top** [reply]
Sorry, MPXV7025DP not the MPX4250, my mistake. 
I set to the MPXV7025DP and it is working.

**Reply by DeeEmm** [reply]
Ahhh good. I was a little confused as I thought that the MPX4250 had been removed some considerable time ago.

**Reply by black-top** [reply]
Yes, sorry. All my ecus uses mpx4250 so it is just stuck in my head and I just selected it from the list.

**Reply by DeeEmm** [reply]
All good. 👍 

Yes it should not have been in the list. (my mistake)

**Reply by black-top** [reply]
Back to the calibration topic. I set maf trim pot at -0.03 and now I see 1.00V at mimic page, but it shows the MAF flow 11.71 kg/h it should be 0. 


> > Just to double check for the calibration, the BOSCH_0280218067 0 to 1V measures reverse flow and above 1V direct flow, with connected maf and no flow I see 1.02V and 11.83kgh flow, so I should adjust the trim pot for 0.02V offset, or adjust until I get 0 flow value, but in that case reverse flow measurement will be ignored?
> 
> The BOSCH_0280218067 does include the reverse flow, so it should be putting out 1v at rest. You will need to adjust for 1v value.

**Reply by DeeEmm** [reply]
> Back to the calibration topic. I set maf trim pot at -0.03 and now I see 1.00V at mimic page, but it shows the MAF flow 11.71 kg/h it should be 0.
> 
> > > Just to double check for the calibration, the BOSCH_0280218067 0 to 1V measures reverse flow and above 1V direct flow, with connected maf and no flow I see 1.02V and 11.83kgh flow, so I should adjust the trim pot for 0.02V offset, or adjust until I get 0 flow value, but in that case reverse flow measurement will be ignored?
> > 
> > 
> > The BOSCH_0280218067 does include the reverse flow, so it should be putting out 1v at rest. You will need to adjust for 1v value.

Yes I can see this in the data....

<img width="340" alt="Screenshot 2025-02-05 at 11 30 22 pm" src="https://github.com/user-attachments/assets/22a7779a-1051-4f95-bd3f-e25612d47f88" />

The right column is the calculated value. It appears to be struggling to deal with the parabolic curve and so is not very accurate around the zero value. If I change the dataset to start from 0cfm this inaccuracy disappears.

**Reply by DeeEmm** [reply]
We may need to just use the forward flow values.

**Reply by black-top** [reply]
Probably yes,  maybe reverse flow for the future.

**Reply by DeeEmm** [reply]
I've replied to this in the MAF sensor thread for the Bosch  BOSCH_0280218067 - https://github.com/DeeEmm/DIY-Flow-Bench/discussions/138#discussioncomment-12075725

**Comment by black-top** [comment]
Small issue, Datagraph settings do not save. Page get reloaded and the settings back to default.
![issue2](https://github.com/user-attachments/assets/eec40567-7963-4bed-9bf7-24082fa78440)

**Reply by DeeEmm** [reply]
I will check

**Reply by black-top** [reply]
Not sure if it is important, but I set the p-ref to mpx4250 and I see at info page p-ref not enabled. 
![issue3](https://github.com/user-attachments/assets/381c3b24-6c45-4d34-be29-a06ecf1cdbd9)

**Reply by DeeEmm** [reply]
will  look at that too. 👍

**Reply by black-top** [reply]
The depression value is stuck at 4.01"h2o, so probably that is causing the issue.
If the mpx4250 sensor is selected the depression value is stuck at 4.01" if I select mpxv7007, I see depression value reacting to the pressure but of course the value is wrong because of different sensor selected.

**Reply by DeeEmm** [reply]
Datagraph datatype value now fixed in both DEV and Master

**Reply by black-top** [reply]
Datagraph setting selection is working now. 
Small improvement could be to add text on the flow graph what correction is used for capturing data. 
Capture all data mode is planned for v3?

**Reply by DeeEmm** [reply]
Yes I think was you before that suggested a data-grid with all info on it?? Was planning to do something along those lines in V3

The capture mode is a good idea too. Will add that to the request.

https://github.com/users/DeeEmm/projects/4?pane=issue&itemId=21297250

**Reply by black-top** [reply]
Yes, we have talked about it some time ago.

**Comment by DeeEmm** [comment]
Added VDO 5WK96132Z MAF (BMW M54B30) and Bosch 0280218038 MAF (Audi S4)

**Reply by black-top** [reply]
Thanks, I have 5WK96132Z MAF it is not a removable insert type but the housing diameter quote similar to 0280218067, so I think I will make adapters to put one maf after the other and will test how it compares. Now when it is possible to select maf type at settings is much easier and faster to do testing like this.

**Reply by DeeEmm** [reply]
I will have a ponder about how we can run multiple MAFs. 

I think that now I've converted MAF processing into a seperate class we can have multiple instances. So theoretically we could hook a MAF up to each of the ADC inputs and run up to four MAF sensors provided that the code caters for the MAF selection / assignment. 

Would also be good to think about how we can support that from a hardware perspective too as currently the ADC channels are already assigned on the shield, although I think it would be possible to jumper from the pressure sensor solder pads to the prototype area / spare edge connector pin.

I think that the existing shield + code could easily be hacked to do this, and not too much more work to tidy it up and integrate it properly

**Reply by black-top** [reply]
Well maybe we can switch pins at settings page, we can select different sensors for same inputs?
I have bought MPXV7007DP for the pitot, and tested today, looks like it is working good, and I use MPXV7025DP for pref,
 so Pdiff input is unused for me, theoretically I could wire up second maf to that input.
The next question what is the best way for displays measured data, to switch between mafs at setings or maybe add a second channel?

**Reply by DeeEmm** [reply]
> Well maybe we can switch pins at settings page, we can select different sensors for same inputs?

Not with the current software, but it could be changed.  As you mention pDiff is not used on a MAF bench and so is essentially 'spare'. 

Anything is possible. Just comes down to how much work / how worthwhile it is. I think that there is a good use case for a dual MAF setup

**Reply by black-top** [reply]
For a current version of the code I'm thinking just use the switch to select between maf sensors and select the correct maf in the settings.

**Reply by DeeEmm** [reply]
... that is two MAF sensors displayed simultaneously
... or two MAF sensors displayed int he same location but selected in accordance with predefined thresholds

e.g. reverse flow - sensor is shown depending on flow direction
e.g. high / low flow - sensor is shown depending on flow value
e.g. both sensors shown in case of using known MAF to map unknown MAF

There's a few different variations but adding some flexibility to account for different use cases is definitely possible.

Can also do crazy things like performing mathematical functions on flow values - For example using two MAF sensors to compare flows in siamesed ports

**Reply by black-top** [reply]
Small issue that flow offset calibration should be 0, to not affect readings for different maf. For multi maf system version probably flow offset calibration will be necessary or each maf.

**Reply by DeeEmm** [reply]
Yes the additional MAF would need a complete additional set of offsets / calibration data.

**Comment by black-top** [comment]
Since we are using CFM, sometimes it is handy to have fps instead m/s for pitot probing, so it will be good to have the option to choose between m/s and fps.

**Reply by DeeEmm** [reply]
> Since we are using CFM, sometimes it is handy to have fps instead m/s for pitot probing, so it will be good to have the option to choose between m/s and fps.

Yes this is a good idea. I will add it to the todo list. 😄

**Comment by DeeEmm** [comment]
OK. I have packaged up RC9 and created the RC9 release.

Official discussion thread - https://github.com/DeeEmm/DIY-Flow-Bench/discussions/367

I have also updated the documentation and the Demo.

If you find any issues, please raise a bug report or comment on discussion linked to above

Hopefully there's no more bugs and we can move on to the official V2 release 🤞 

Next I will put together a poll with items for the V3 release. There are some items that will automatically get included, but it is a chance for everyone to vote on what they want to see in the next planned release.

**Reply by DeeEmm** [reply]
POLL - https://github.com/DeeEmm/DIY-Flow-Bench/discussions/368
